source/common/mongo/proxy.cc

```diff
@@ -241,5 +245,54 @@ DecoderPtr ProdProxyFilter::createDecoder(DecoderCallbacks& callbacks) {
   return DecoderPtr{new DecoderImpl(callbacks)};
 }
 
+FaultConfigImpl::FaultConfigImpl(const Json::Object& fault_config)
+    : delay_percent_(
+          static_cast<uint32_t>(fault_config.getObject("fixed_delay")->getInteger("percent"))),
+      duration_ms_(
+          static_cast<uint64_t>(fault_config.getObject("fixed_delay")->getInteger("duration_ms"))) {
+}
+
+Optional<uint64_t> ProxyFilter::delayDuration() {
+  Optional<uint64_t> result;
+
+  if (!fault_config_)
+    return result;
+
+  if (!runtime_.snapshot().featureEnabled("mongo.fault.delay.percent",
+                                          fault_config_->delayPercent())) {
+    return result;
+  }
+
+  uint64_t duration = runtime_.snapshot().getInteger("mongo.fault.delay.duration_ms",
+                                                     fault_config_->delayDuration());
+
+  // Delay only if the duration is > 0ms.
+  if (duration > 0) {
+    result.value(duration);
+  }
+
+  return result;
+}
+
+void ProxyFilter::delayInjectionTimerCallback() {
```

Comment:
this isn't the callback that you want. We want a delay to operate on an active query. When the timer trigggers this code is telling the filter to continue on reading instead of processing the activequery that the delay should be associated it.

Reply:
adding handling of this situation.

Reply:
this is fixed in the next iteration.

Reply:
this needs to be fixed to capture stats via decode. I'll fix it today.